### ISSUE TITLE
PeekableReadBuffer: Memory limit exceed

### ISSUE THREAD

**Issue by liujiawinds** [ClickHouse/ClickHouse#17300]
I encountered an exception when insert a big column value into table. It looks like buffer size is not enough. But I'm wondering which configuration item can adjust it.

Thanks in advance.

```
2020.11.23 18:20:16.192024 [ 75806 ] {bbe2f516-5c08-4965-8a2f-b65cc71042e4} <Error> DynamicQueryHandler: Code: 241, e.displayText() = DB::Exception: PeekableReadBuffer: Memory limit exceed, Stack trace (when copying this message, always include the lines below):

0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x18e1b360 in /usr/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xe736dad in /usr/bin/clickhouse
2. ? @ 0x165fbe82 in /usr/bin/clickhouse
3. DB::PeekableReadBuffer::peekNext() @ 0x165fb1d3 in /usr/bin/clickhouse
4. DB::PeekableReadBuffer::nextImpl() @ 0x165fb518 in /usr/bin/clickhouse
5. DB::ValuesBlockInputFormat::skipToNextRow(unsigned long, int) @ 0x166d2cfd in /usr/bin/clickhouse
6. DB::ValuesBlockInputFormat::parseExpression(DB::IColumn&, unsigned long) @ 0x166d33cc in /usr/bin/clickhouse
7. ? @ 0x166d75d7 in /usr/bin/clickhouse
8. DB::ValuesBlockInputFormat::generate() @ 0x166d5e6d in /usr/bin/clickhouse
9. DB::ISource::work() @ 0x1651f65b in /usr/bin/clickhouse
10. DB::InputStreamFromInputFormat::readImpl() @ 0x164e364d in /usr/bin/clickhouse
11. DB::IBlockInputStream::read() @ 0x156c6d3d in /usr/bin/clickhouse
12. DB::InputStreamFromASTInsertQuery::readImpl() @ 0x15acc749 in /usr/bin/clickhouse
13. DB::IBlockInputStream::read() @ 0x156c6d3d in /usr/bin/clickhouse
14. DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::__1::atomic<bool>*) @ 0x156e9cce in /usr/bin/clickhouse
15. DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, bool, DB::Context&, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)>) @ 0x15e31c65 in /usr/bin/clickhouse
16. DB::HTTPHandler::processQuery(DB::Context&, Poco::Net::HTTPServerRequest&, HTMLForm&, Poco::Net::HTTPServerResponse&, DB::HTTPHandler::Output&) @ 0x1647cdd9 in /usr/bin/clickhouse
17. DB::HTTPHandler::handleRequest(Poco::Net::HTTPServerRequest&, Poco::Net::HTTPServerResponse&) @ 0x1648088b in /usr/bin/clickhouse
18. Poco::Net::HTTPServerConnection::run() @ 0x18cfafb3 in /usr/bin/clickhouse
19. Poco::Net::TCPServerConnection::start() @ 0x18d391cb in /usr/bin/clickhouse
20. Poco::Net::TCPServerDispatcher::run() @ 0x18d3965b in /usr/bin/clickhouse
21. Poco::PooledThread::run() @ 0x18eb8136 in /usr/bin/clickhouse
22. Poco::ThreadImpl::runnableEntry(void*) @ 0x18eb3530 in /usr/bin/clickhouse
23. start_thread @ 0x7ea5 in /usr/lib64/libpthread-2.17.so
24. clone @ 0xfe8dd in /usr/lib64/libc-2.17.so
 (version 20.9.3.45 (official build))
```

**Comment by den-crane** [comment]
try to disable input_format_parallel_parsing

--input_format_parallel_parsing arg                              Enable parallel parsing for some data formats.

**Comment by liujiawinds** [comment]
@den-crane It seems doesn't work, the same error message occured after I inserted the configuration item into `users.xml` and restart clickhouse-server process. 
![image](https://user-images.githubusercontent.com/1757754/100035416-1b3e1880-2e39-11eb-82f5-83df33c440d7.png)

**Comment by den-crane** [comment]
try 0

```
<input_format_parallel_parsing>0</input_format_parallel_parsing>
```

restart is not needed.

**Comment by filimonov** [comment]
>  It looks like buffer size is not enough. 

Buffer tries to grow to fit your fat column automatically. But during that attempt it read the memory limitation.

**Comment by liujiawinds** [comment]
@den-crane That not work too. 
I found a work around that use binary stream api solve this problem. 

```
long[] longs = termList.stream().flatMap(Collection::stream).mapToLong(v -> Long.valueOf(v.toString())).toArray();
        connection.createStatement().sendRowBinaryStream(
                "insert into bloom_test10 (file_name, original_log_bloom)",
                stream -> {
                    stream.writeString(path);
                    stream.writeUInt32Array(longs);
                }
        );
```

**Comment by astudnev** [comment]
Please reopen this, as i getting the same errors after updated from 19.9 to 20.11

```
2021.01.03 08:01:15.601198 [ 1968 ] {b4472a5e-e463-4852-a0ed-efd90fae6393} <Error> DynamicQueryHandler: Code: 241, e.displayText() = DB::Exception: PeekableReadBuffer: Memory limit exceed, Stack trace (when copying this message, always include the lines below):

0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x18cd2050 in /usr/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xe63232d in /usr/bin/clickhouse
2. ? @ 0x164be2e2 in /usr/bin/clickhouse
3. DB::PeekableReadBuffer::peekNext() @ 0x164bd633 in /usr/bin/clickhouse
4. DB::PeekableReadBuffer::nextImpl() @ 0x164bd978 in /usr/bin/clickhouse
5. void DB::readQuotedStringInto<true, DB::PODArray<char8_t, 4096ul, Allocator<false, false>, 15ul, 16ul> >(DB::PODArray<char8_t, 4096ul, Allocator<false, false>, 15ul, 16ul>&, DB::ReadBuffer&) @ 0xe67e4f5 in /usr/bin/clickhouse
6. DB::DataTypeString::deserializeTextQuoted(DB::IColumn&, DB::ReadBuffer&, DB::FormatSettings const&) const @ 0x156445a4 in /usr/bin/clickhouse
7. DB::ValuesBlockInputFormat::readRow(std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >&, unsigned long) @ 0x16597b4b in /usr/bin/clickhouse
8. DB::ValuesBlockInputFormat::generate() @ 0x165982cd in /usr/bin/clickhouse
9. DB::ISource::work() @ 0x163e821b in /usr/bin/clickhouse
10. DB::InputStreamFromInputFormat::readImpl() @ 0x163ab46d in /usr/bin/clickhouse
11. DB::IBlockInputStream::read() @ 0x155611cd in /usr/bin/clickhouse
12. DB::InputStreamFromASTInsertQuery::readImpl() @ 0x15973f79 in /usr/bin/clickhouse
13. DB::IBlockInputStream::read() @ 0x155611cd in /usr/bin/clickhouse
14. DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::__1::atomic<bool>*) @ 0x1558415e in /usr/bin/clickhouse
15. DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, bool, DB::Context&, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)>) @ 0x15d06285 in /usr/bin/clickhouse
16. DB::HTTPHandler::processQuery(DB::Context&, Poco::Net::HTTPServerRequest&, HTMLForm&, Poco::Net::HTTPServerResponse&, DB::HTTPHandler::Output&) @ 0x16344c69 in /usr/bin/clickhouse
17. DB::HTTPHandler::handleRequest(Poco::Net::HTTPServerRequest&, Poco::Net::HTTPServerResponse&) @ 0x1634871b in /usr/bin/clickhouse
18. Poco::Net::HTTPServerConnection::run() @ 0x18bb1df3 in /usr/bin/clickhouse
19. Poco::Net::TCPServerConnection::start() @ 0x18bf000b in /usr/bin/clickhouse
20. Poco::Net::TCPServerDispatcher::run() @ 0x18bf0728 in /usr/bin/clickhouse
21. Poco::PooledThread::run() @ 0x18d6ee36 in /usr/bin/clickhouse
22. Poco::ThreadImpl::runnableEntry(void*) @ 0x18d6a230 in /usr/bin/clickhouse
23. start_thread @ 0x7494 in /lib/x86_64-linux-gnu/libpthread-2.24.so
24. clone @ 0xe8aff in /lib/x86_64-linux-gnu/libc-2.24.so
 (version 20.8.11.17 (official build))
```

Settings 
<input_format_parallel_parsing>0</input_format_parallel_parsing>
and
<input_format_values_interpret_expressions>0</input_format_values_interpret_expressions>

do not help

**Comment by filimonov** [comment]
See https://github.com/ClickHouse/ClickHouse/issues/18690

Fixed in 21.1 and newer: https://github.com/ClickHouse/ClickHouse/pull/18979
